### PR TITLE
Enh/base - Initial framework written

### PR DESCRIPTION
**Pull request by Besler**
@araex @Yves Could you review this?

The basic premise is this:
`MultiScaleHessianEnhancementImageFilter` is the work horse. You provide a set of single values through `MultiScaleHessianEnhancementImageFilter.SetSigmaArray()`. Internally, this calculated the hessian at each voxel, computes the eigenvalue at each voxel, converts the eigenvalue to a scalar measures, and takes maximum absolute value response across all sigmas using `MaximumAbsoluteValueImageFilter`.

To convert the eigenvalues to a scalar measure, a class that inherits from `EigenvalueToScalarImageFilter` is used. You create that filter outside the class and set it (similar to registration) using `SetEigenToScalarImageFilter( )`.

The next step is to implement all these different methods (derived from `SetEigenToScalarImageFilter( )` and their parameter selection methods) into one class.

I wanted to include the previous history from [ITK-KrcahSheetnessImageFilter ](https://github.com/ypauchard/ITK-KrcahSheetnessImageFilter) but it would have been a mess.

Bryce

### CODE REVIEW

**Comment by Besler** [comment]
I haven't figured out the python stuff yet. I will do all that once the C code is working and complete.

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+  /* Populate that array */
+  SigmaType thisSigma;
+  for (SigmaStepsType scaleLevel = 0; scaleLevel < sigmaArray.GetSize(); ++scaleLevel)
+  {
+    switch ( SigmaStepMethod )
+    {
+    case Self::EquispacedSigmaSteps:
+      {
+      const RealType stepSize = std::max( 1e-10, ( SigmaMaximum - SigmaMinimum ) / ( NumberOfSigmaSteps - 1 ) );
```
Comment (araex) [inline include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Division by 0 if NumberOfSigmaSteps is 1

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /* If we have min greater than max just swap */
+  if (SigmaMinimum > SigmaMaximum) {
+    std::swap(SigmaMinimum, SigmaMaximum);
+  }
+
+  /* If min equals max, just generate one step size */
+  if (SigmaMinimum == SigmaMaximum) {
+    NumberOfSigmaSteps = 1;
```
Comment (araex) [inline include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Leads to a division by 0 further down

File: include/itkEigenvalueToScalarImageFilter.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+   * itkSymmetricEigenAnalysisImageFilter.h or itkSymmetricEigenAnalysis.h. That turns out to be non-trivial
+   * because the enumeration is hidden within the templated class. Therefore, you would need the hessian type
+   * and eigenvalue type to do such an operation. We do not necessarily have the hessian type information.
+   * */
+  typedef enum {
+    OrderByValue = 1,
+    OrderByMagnitude,
+    DoNotOrder
+  } EigenValueOrderType;
+  itkStaticConstMacro(EigenValueOrder, EigenValueOrderType, OrderByMagnitude);
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkEigenvalueToScalarImageFilter.h]: Good point! Now there is a pure virtual function GetEigenValueOrder.

File: include/itkEigenvalueToScalarImageFilter.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                   ( Concept::SameDimension< TInputImage::ImageDimension, TOutputImage::ImageDimension >) );
+  // End concept checking
+protected:
+  void PrintSelf(std::ostream & os, Indent indent) const ITK_OVERRIDE {
+    Superclass::PrintSelf(os, indent);
+    os << indent << "EigenValueOrder: " << this->GetEigenValueOrder() << std::endl;
+  }
+
+private:
+  ITK_DISALLOW_COPY_AND_ASSIGN(EigenvalueToScalarImageFilter);
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkEigenvalueToScalarImageFilter.h]: Right now, the compiler would create a default constructor which is the issue, correct? Now that GetEigenValueOrder is a pure virtual function the class cannot be instantiated.

File: include/itkMultiScaleHessianEnhancementImageFilter.h [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    LogarithmicSigmaSteps = 1
+  } SigmaStepMethodEnum;
+
+  /** Set/Get macros for SigmaArray */
+  itkSetMacro(SigmaArray, SigmaType);
+  itkGetConstMacro(SigmaArray, SigmaType);
+
+  /**
+   * Static methods for generating an array of sigma values. Note that these still need to be passed
+   * into the class using TODO: link to SetSigmaArray. Implementation taken from itkMultiScaleHessianBasedMeasureImageFilter.
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.h]: Fixed! See c298122d9bd7911118d650e34715722bf7dc414d

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    float numberOfFiltersToProcess = 3*m_SigmaArray.GetSize() + 1*(m_SigmaArray.GetSize()-1);
+    float perFilterProccessPercentage = 1.0 / numberOfFiltersToProcess;
+    itkDebugMacro(<< "each filter accounts for " << perFilterProccessPercentage*100.0 << "% of processing");
+
+    progress->RegisterInternalFilter(m_HessianFilter, perFilterProccessPercentage);
+    progress->RegisterInternalFilter(m_EigenAnalysisFilter, perFilterProccessPercentage);
+    progress->RegisterInternalFilter(m_EigenToScalarImageFilter, perFilterProccessPercentage);
+
+    /* Check if we need to run the MaximumAbsoluteValueFilter at all */ 
+    if (m_SigmaArray.GetSize() > 1)
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Yep, no idea what was going on there. See a3fed8f98f91a9b1ecd7cac98c3fc954c2467210

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    {
+      itkDebugMacro(<< "maximumAbsoluteValueFilter is not being used");
+    }
+  }
+
+  /* We store a single pointer that we will graft to the output */
+  typename TOutputImage::Pointer outputImageTypePointer;
+
+  /* Process the first scale */
+  outputImageTypePointer = generateResponseAtScale((SigmaType)0);
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Good catch! See 261c4299ec9aadd4e5a829b75390830199d45e5f

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+template< typename TInputImage, typename TOutputImage >
+typename MultiScaleHessianEnhancementImageFilter< TInputImage, TOutputImage >::SigmaArrayType
+MultiScaleHessianEnhancementImageFilter< TInputImage, TOutputImage >
+::GenerateSigmaArray(SigmaType SigmaMinimum, SigmaType SigmaMaximum, SigmaStepsType NumberOfSigmaSteps, SigmaStepMethodEnum SigmaStepMethod)
+{
+  /* Quick check to make sure value is correct */
+  if ( NumberOfSigmaSteps < 1 )
+  {
+    /*
+    * Since this is a static class, we have no "this" pointer so we cannot throw an exception. Instead
```
[the remark this answers is not in the document]
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Perfect! See 0ed3cf60ff51ccfa082da899bddcb7dc1df1a0a0

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx (hunk 2 above)
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Good find. Strange that it passed the tests... See ce229e2d94ba40f90588298a8e0db649bf84e491

File: include/itkMultiScaleHessianEnhancementImageFilter.hxx (hunk 1 above)
Comment (Besler) [reply include/itkMultiScaleHessianEnhancementImageFilter.hxx]: Good find. Strange that it passed the tests... See ce229e2d94ba40f90588298a8e0db649bf84e491

**Pull request merged**